### ISSUE TITLE
HeapCache - putAll method gives NullPointerException when checking for null

### ISSUE THREAD

**Issue by mnouh** [cache2k/cache2k#90]
Issue is HeapCache.java, specifically the putAll method where it checks for null keys.
The underlying valueMap can be a ConcurrentHashMap that throws a NullPointerException when checking for null keys. ConcurrentHashMap does not allow for checking for null keys/values.

    /**
     * Returns <tt>true</tt> if this map contains a mapping for the specified
     * key.  More formally, returns <tt>true</tt> if and only if
     * this map contains a mapping for a key <tt>k</tt> such that
     * <tt>(key==null ? k==null : key.equals(k))</tt>.  (There can be
     * at most one such mapping.)
     *
     * @param key key whose presence in this map is to be tested
     * @return <tt>true</tt> if this map contains a mapping for the specified
     *         key
     * @throws ClassCastException if the key is of an inappropriate type for
     *         this map
     * (<a href="{@docRoot}/java/util/Collection.html#optional-restrictions">optional</a>)
     * @throws NullPointerException if the specified key is null and this map
     *         does not permit null keys
     * (<a href="{@docRoot}/java/util/Collection.html#optional-restrictions">optional</a>)
     */
    boolean containsKey(Object key);

```
public void putAll(Map<? extends K, ? extends V> valueMap) {
        if (valueMap.containsKey((Object)null)) {
            throw new NullPointerException("map contains null key");
        } else {
            Iterator var2 = valueMap.entrySet().iterator();

            while(var2.hasNext()) {
                java.util.Map.Entry<? extends K, ? extends V> e = (java.util.Map.Entry)var2.next();
                this.put(e.getKey(), e.getValue());
            }

        }
    }
```

**Comment by cruftex** [comment]
Ahh, I am wrong its actually in the contract of `Map` the hopping point is:

> @throws `NullPointerException` if the specified key is null and this map
>         does not permit `null` keys

**Comment by cruftex** [comment]
I will fix it in the upcoming 1.2 release tomorrow.
